go/chat/flipmanager.go

```diff
@@ -0,0 +1,736 @@
+	m.gamesMu.Lock()
+	defer m.gamesMu.Unlock()
+	if len(m.dirtyGames) == 0 {
+		return
+	}
+	defer func() {
+		m.dirtyGames = make(map[string]chat1.FlipGameID)
+	}()
+	ctx := m.makeBkgContext()
+	ui, err := m.G().UIRouter.GetChatUI()
+	if err != nil || ui == nil {
+		m.Debug(ctx, "notifyDirtyGames: no chat UI available for notification")
+		return
+	}
+	var updates []chat1.UICoinFlipStatus
+	for _, dg := range m.dirtyGames {
+		if game, ok := m.games.Get(dg.String()); ok {
+			updates = append(updates, game.(chat1.UICoinFlipStatus))
+		}
+	}
+	if err := ui.ChatCoinFlipStatus(ctx, updates); err != nil {
+		m.Debug(ctx, "notifyDirtyGames: failed to notify status: %s", err)
+	}
+}
+
+func (m *FlipManager) notificationLoop(shutdownCh chan struct{}) {
+	duration := 200 * time.Millisecond
+	next := m.clock.Now().Add(duration)
+	for {
+		select {
+		case <-m.clock.AfterTime(next):
+			m.notifyDirtyGames()
+			next = m.clock.Now().Add(duration)
+		case <-m.forceCh:
+			m.notifyDirtyGames()
+			next = m.clock.Now().Add(duration)
+		case <-shutdownCh:
+			return
+		}
+	}
+}
+
+func (m *FlipManager) addParticipant(ctx context.Context, status *chat1.UICoinFlipStatus,
+	update flip.CommitmentUpdate) {
+	username, deviceName, _, err := m.G().GetUPAKLoader().LookupUsernameAndDevice(ctx,
+		keybase1.UID(update.User.U.String()), keybase1.DeviceID(update.User.D.String()))
+	if err != nil {
+		m.Debug(ctx, "addParticipant: failed to get username/device (using IDs): %s", err)
+		username = libkb.NewNormalizedUsername(update.User.U.String())
+		deviceName = update.User.D.String()
+	}
+	status.Participants = append(status.Participants, chat1.UICoinFlipParticipant{
+		Uid:        update.User.U.String(),
+		DeviceID:   update.User.D.String(),
+		Username:   username.String(),
+		DeviceName: deviceName,
+		Commitment: update.Commitment.String(),
+	})
+	sort.Slice(status.Participants, func(i, j int) bool {
+		return status.Participants[i].Username < status.Participants[j].Username
+	})
+	endingS := ""
+	if len(status.Participants) > 1 {
+		endingS = "s"
+	}
+	status.ProgressText = fmt.Sprintf("Gathered %d commitment%s", len(status.Participants), endingS)
+}
+
+func (m *FlipManager) addReveal(ctx context.Context, status *chat1.UICoinFlipStatus,
+	update flip.RevealUpdate) {
+	numReveals := 0
+	for index, p := range status.Participants {
+		if p.Reveal != nil {
+			numReveals++
+		}
+		if p.Uid == update.User.U.String() && p.DeviceID == update.User.D.String() {
+			reveal := update.Reveal.String()
+			status.Participants[index].Reveal = &reveal
+			numReveals++
+		}
+	}
+	status.ProgressText = fmt.Sprintf("%d participants have revealed secrets", numReveals)
+}
+
+func (m *FlipManager) addResult(ctx context.Context, status *chat1.UICoinFlipStatus, result flip.Result,
+	convID chat1.ConversationID) {
+	defer func() {
+		if len(status.ResultText) > 0 {
+			status.ProgressText += " (complete)"
+		}
+	}()
+	hmi, err := m.getHostMessageInfo(ctx, convID)
+	if err != nil {
```

Comment:
might be slightly better as a `switch`, no?

Reply:
Wasn't too sure how to make this a switch.

Reply:
```switch {
   case foo != nil:
   case bar != nil:
}```

no?

Reply:
https://tour.golang.org/flowcontrol/11

Reply:
TIL